### PR TITLE
Change RA mechanic to repair ally-owned husks

### LINKED ISSUE #9960: Idea for mechanics and ally husks
Was playing RA multilayer the other day and one of my allies lost their MCV, I put a mechanic in to try to revive it, however, it came back as my unit, would be good if it remained their unit if the husk belongs to an ally. There have also been cases where I've wanted to allow an ally to capture my second construction yard, could be good to allow allies buildings to be captured, not sure whether this would be better if you explicitly allowed it, not doing so would open the game up to ally sabotage, which would take a lot of collaborative planning so might be worth allowing as a game tactic?
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28144177-idea-for-mechanics-and-ally-husks?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #14019: Bug report : repair burned friend mcv and get friend unit abilities
in multiplayer 2v2 game one of enemies build mcv and send that to his friend base
allied person explode friend soviet mcv and repair that with mechanic and get new mcv with soviet abilities!!!
after that allied person can build IC and Tesla and soviet building & units & superpowers !!!!
its really bad bug and after i see this in game i surrend because i cant win this unfare game!!!
replay attached here  : 
[OpenRA-2017-09-12T151523Z.orarep.zip](https://github.com/OpenRA/OpenRA/files/1299820/OpenRA-2017-09-12T151523Z.orarep.zip)

### PR DESCRIPTION
**Pull request by pchote**
Fixes #14019.
Fixes #9960.

Mechanics now show the goldwrench cursor for own- and ally-husks, and restore vehicles to their original owner.  Capturing vs enemy-husks has not changed, and continues to use the "enter" cursor.  This is close enough to the way engineers handle repair vs capturing that it should hopefully not cause confusion.

### CODE REVIEW

**Review by Smittytron — DISMISSED** [review]
Works as advertised.

I'd say this closes #9960. The building part of that ticket has a duplicate issue (#5691).

**Comment by KOYK** [comment]
@pchote did you add the ability to disable\enable it in the lobby as an option?

File: OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+using System;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("This structure can be infiltrated causing funds to be stolen.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs]: This `Desc` doesn't look correct.

File: OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	class InfiltrateForTransformInfo : ITraitInfo
+	{
+		[ActorReference, FieldLoader.Require] public readonly string IntoActor = null;
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs]: Style nit: `public readonly string IntoActor = null;` on its own line.

File: OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		}
+
+		void INotifyInfiltrated.Infiltrated(Actor self, Actor infiltrator, BitSet<TargetableType> types)
+		{
+			if (!info.Types.Overlaps(types))
+				return;
+
+			var facing = self.TraitOrDefault<IFacing>();
+			var transform = new Transform(self, info.IntoActor) { ForceHealthPercentage = info.ForceHealthPercentage, Faction = faction };
+			if (facing != null) transform.Facing = facing.Facing;
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs]: Style nit: `transform.Facing = facing.Facing;` on its own line with an empty line afterwards, please.

File: OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		[ActorReference, FieldLoader.Require] public readonly string IntoActor = null;
+		public readonly int ForceHealthPercentage = 0;
+		public readonly bool SkipMakeAnims = true;
+
+		public readonly BitSet<TargetableType> Types = default(BitSet<TargetableType>);
+
+		public object Create(ActorInitializer init) { return new InfiltrateForTransform(init, this); }
+	}
+
+	class InfiltrateForTransform : INotifyInfiltrated
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs]: Can we name this `TransformOnInfiltration` to stay in line with `TransformOnCapture`?

File: OpenRA.Mods.Cnc/Traits/Infiltration/Infiltrates.cs [hunk 5]
```diff
@@ -35,12 +36,14 @@ public class InfiltratesInfo : ConditionalTraitInfo
 		public readonly EnterBehaviour EnterBehaviour = EnterBehaviour.Dispose;
 
 		[NotificationReference("Speech")]
-		[Desc("Notification to play when a building is infiltrated.")]
-		public readonly string Notification = "BuildingInfiltrated";
+		[Desc("Notification to play when infiltrated.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Infiltration/Infiltrates.cs]: `infiltrated` -> `infiltrating`. Spies are not infiltrated.

**Review by abcdefg30 — COMMENTED** [review]
This will effectively revert https://github.com/OpenRA/OpenRA/pull/13276#issuecomment-338411432.

Line 34 of `Infiltrates.cs` needs an adjustment as well.

**Comment by pchote** [comment]
> This will effectively revert https://github.com/OpenRA/OpenRA/pull/13276#issuecomment-338411432.

This was superseded by #14982

File: OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs (hunk 4 above)
Comment (pchote) [reply OpenRA.Mods.Cnc/Traits/Infiltration/InfiltrateForTransform.cs]: This would make it inconsistent with all the other `InfiltrateFor*` traits.

**Comment by pchote** [comment]
Fixed.

**Review by obrakmann — APPROVED** [review]

**Pull request merged**